dynamic/json.go

```diff
@@ -56,18 +59,34 @@ func (m *Message) marshalJSON(b *indentBuffer) error {
 		return err
 	}
 
+	emitDefaults := opts.EmitDefaults
+
+	var tags []int
+	if emitDefaults {
+		tags = m.allKnownFieldTags()
+	} else {
+		tags = m.knownFieldTags()
+	}
+
 	first := true
+
 	// first the known fields
-	for _, tag := range m.knownFieldTags() {
+	for _, tag := range tags {
 		itag := int32(tag)
-		v := m.values[itag]
 		fd := m.FindFieldDescriptor(itag)
 
+		v, ok := m.values[itag]
+		if !ok {
+			if emitDefaults {
```

Comment:
I don't think you need to double-check this flag: we should only ever get here if `emitDefaults`. (Also means you can remove the `emitDefaults` local variable and inline into its one other use.)

It may be possible to incorrectly reach this line through some concurrency bug in calling code, but I don' think we want to go out of our way to handle that (and, if we did, the right way to address it would probably be to `continue` if `!emitDefaults`...)